File: backend/app/routers/buds.py
from __future__ import annotations
from fastapi import APIRouter, Depends, HTTPException
from supabase import Client

from app.deps import get_db, require_user
from app.repositories.bud_repo import BudRepository
from app.schemas.bud import BudHistoryOut, BudMoveRequest, BudOut, BudPatch, BudProgressUpdate
from app.services.bud_service import BudService
from app import undo_store

router = APIRouter(prefix="/buds", tags=["buds"])


@router.get("")
def list_buds(plant_id: str | None = None, wilting_only: bool = False,
              user=Depends(require_user), db: Client = Depends(get_db)):
    buds = BudService(db).list(user.id, plant_id=plant_id, wilting_only=wilting_only)
    return {"ok": True, "data": {"items": [BudOut.model_validate(b) for b in buds]}}


@router.get("/{bud_id}")
def get_bud(bud_id: str, user=Depends(require_user), db: Client = Depends(get_db)):
    svc = BudService(db)
    bud, history = svc.get_with_history(user.id, bud_id)
    if not bud:
        raise HTTPException(404, "봉우리를 찾을 수 없습니다.")
    return {
        "ok": True,
        "data": {
            "bud": BudOut.model_validate(bud),
            "history": [BudHistoryOut.model_validate(h) for h in history],
        },
    }


@router.patch("/{bud_id}")
def patch_bud(bud_id: str, body: BudPatch, user=Depends(require_user), db: Client = Depends(get_db)):
    svc = BudService(db)
    bud = svc.get(user.id, bud_id)
    if not bud:
        raise HTTPException(404)
    updated = BudRepository(db).update(user.id, bud_id, body.model_dump(exclude_none=True))
    return {"ok": True, "data": BudOut.model_validate(updated or bud)}


@router.patch("/{bud_id}/move")
def move_bud(bud_id: str, body: BudMoveRequest, user=Depends(require_user), db: Client = Depends(get_db)):
    try:
        updated = BudService(db).move_to_plant(user.id, bud_id, body.target_plant_id)
    except ValueError as exc:
        raise HTTPException(404, str(exc)) from exc
    return {"ok": True, "data": BudOut.model_validate(updated)}

@router.delete("/{bud_id}")
def delete_bud(bud_id: str, user=Depends(require_user), db: Client = Depends(get_db)):
    try:
        bud = BudService(db).get(user.id, bud_id)
        BudService(db).delete(user.id, bud_id)
        undo_store.push(user.id, "bud_restore", f"봉우리 '{bud.title}' 삭제", vars(bud))
    except ValueError as exc:
        raise HTTPException(404, str(exc)) from exc
    return {"ok": True, "data": {"deleted_id": bud_id}}


@router.patch("/{bud_id}/progress")
def set_bud_progress(bud_id: str, body: BudProgressUpdate,
                     user=Depends(require_user), db: Client = Depends(get_db)):
    """Directly set a bud's progress (manual slider). Applies the same
    auto status-transition + history as the AI skill. The optional note is the
    user's reason for the change."""
    svc = BudService(db)
    bud = svc.get(user.id, bud_id)
    if not bud:
        raise HTTPException(404, "봉우리를 찾을 수 없습니다.")
    note = body.note.strip() or "사용자 수동 조정"
    updated = svc.update_progress(user.id, bud_id, body.progress, auto_transition=True, note=note)
    return {"ok": True, "data": BudOut.model_validate(updated or bud)}

File: frontend/lib/api/buds.ts
import { apiDelete, apiGet, apiPatch } from "./client";

export const listBuds = (params: { plant_id?: string; wilting_only?: boolean } = {}) => {
  const q = new URLSearchParams();
  if (params.plant_id) q.set("plant_id", params.plant_id);
  if (params.wilting_only) q.set("wilting_only", "true");
  return apiGet<{ items: Bud[] }>(`/buds?${q}`);
};

export const getBud = (id: string) =>
  apiGet<{ bud: Bud; history: BudHistory[] }>(`/buds/${id}`);

export const patchBud = (id: string, fields: { title?: string; detail?: string }) =>
  apiPatch<Bud>(`/buds/${id}`, fields);

export const deleteBud = (id: string) =>
  apiDelete<{ deleted_id: string }>(`/buds/${id}`);

/** Manually set a bud's progress (slider). `note` carries the user's reason. */
export const setBudProgress = (id: string, progress: number, note = "") =>
  apiPatch<Bud>(`/buds/${id}/progress`, { progress, note });

export const moveBud = (id: string, targetPlantId: string) =>
  apiPatch<Bud>(`/buds/${id}/move`, { target_plant_id: targetPlantId });

export interface Bud {
  id: string;
  plant_id: string;
  title: string;
  detail: string;
  type: "concern" | "schedule";
  status: "seed" | "bud" | "flower" | "fruit" | "harvested" | "wilting" | "rot"; // seed: migration 004 이전 읽기 호환
  progress: number;
  deadline: string | null;
  last_progress_at: string | null;
  disappeared_at: string | null;
  created_at: string;
}

export interface BudHistory {
  id: string;
  bud_id: string;
  from_status: string;
  to_status: string;
  at: string;
  reason: string;
}
